[Merged by Bors] - fix(tactic/lint/simp): only head-beta reduce, don't whnf

gebner opened the pull request:
This is necessary to accept simp lemmas like `injective reverse`.

---

See #4222 for motivation.

I don't know if this will break anything.

gebner committed:
Make simp linters only head-beta reduce instead of whnf

gebner committed:
data/list/basic: mark more injectivity lemmas simp

eric-wieser commented on test/lint_simp_var_head.lean:
```diff
@@ -22,3 +22,15 @@ decl ← get_decl ``const_zero_eq_zero,
+
+
+-- However injectivity lemmas can still be marked simp,
+-- even though injective is reducible and unfolds to a bad simp lemma:
```
What do you mean by "bad" here?

gebner replied:
Bad in the sense of the `simp_var_head` linter (which this file tests).  If you unfold `injective f`, then you get `∀ x y, f x = f y → x = y` where the left-hand side `x` is a variable.  This is bad because the simplifier code doesn't support this, and the linter makes sure we don't mark it as a simp lemma.

gebner committed:
Do not mark ne_empty_of_mem as simp.

This lemma can never fire anyhow because of the `a ∈ s` condition, where
`a` does not occur on the lhs.
